npe2/_inspection/_visitors.py

```diff
@@ -0,0 +1,455 @@
+        # for each in the decorator list ...
+        for call in node.decorator_list:
+            # https://docs.python.org/3/library/ast.html#ast.Call
+            if isinstance(call, ast.Name) and self._names.get(call.id, "").startswith(
+                self._match
+            ):
+                self._process_decorated(call.id, node, {})
+
+            if not isinstance(call, ast.Call):
+                continue
+            # if the function is an attribute ...
+            # (e.g in `@npe2.implements.reader`, `reader` is an attribute of
+            # implements, which is an attribute of npe2)
+            if isinstance(call.func, ast.Attribute):
+                # then go up the chain of attributes until we get to a root Name
+                val = call.func.value
+                _names = []
+                while isinstance(val, ast.Attribute):
+                    _names.append(val.attr)
+                    val = val.value
+                if isinstance(val, ast.Name):
+                    _names.append(val.id)
+                    # finally, we can build the fully resolved call name of the
+                    # decorator (e.g. `@npe2.implements`, or `@implements`)
+                    call_name = ".".join(reversed(_names))
+                    # we then check the `_names` we gathered during imports to resolve
+                    # these call names to their fully qualified names (e.g. `implements`
+                    # would resolve to `npe2.implements` if the name `implements` was
+                    # imported from `npe2`.)
+                    # If the name resolves to the name of this module,
+                    # then we have a hit! process the function.
+                    if self._names.get(call_name) == self._match:
+                        kwargs = self._keywords_to_kwargs(call.keywords)
+                        self._process_decorated(call.func.attr, node, kwargs)
+            elif isinstance(call.func, ast.Name):
+                # if the function is just a direct name (e.g. `@reader`)
+                # then we can just see if the name points to something imported from
+                # this module.
+                if self._names.get(call.func.id, "").startswith(self._match):
+                    kwargs = self._keywords_to_kwargs(call.keywords)
+                    self._process_decorated(call.func.id, node, kwargs)
+        return super().generic_visit(node)
+
+    def _keywords_to_kwargs(self, keywords: List[ast.keyword]) -> Dict[str, Any]:
+        return {str(k.arg): ast.literal_eval(k.value) for k in keywords}
+
+    @abstractmethod
+    def _process_decorated(
+        self,
+        decorator_name: str,
+        node: Union[ast.ClassDef, ast.FunctionDef],
+        decorator_kwargs: dict,
+    ):
+        """Process a decorated function.
+
+        This is a hook for subclasses to do something with the decorated function.
+        """
+
+
+class NPE2PluginModuleVisitor(_DecoratorVisitor):
```

Comment:
copied from `npe2/implements` on main, no changes except the class name changed from `PluginModuleVisitor`